internal/controller/volumereplicationgroup_controller.go

```diff
@@ -800,41 +795,267 @@ func (v *VRGInstance) separatePVCsUsingVRGStatus(pvcList *corev1.PersistentVolum
+	total := copy(v.volRepPVCs, pvcList.Items)
+	v.log.Info("Found PersistentVolumeClaims", "count", total)
+
+	return nil
+}
+
+//nolint:gocognit,cyclop
+func (v *VRGInstance) separatePVCsUsingStorageClassProvisioner(pvcList *corev1.PersistentVolumeClaimList) error {
+	peerClasses := v.instance.Spec.Async.PeerClasses
+
+	for idx := range pvcList.Items {
+		pvc := &pvcList.Items[idx]
+		scName := pvc.Spec.StorageClassName
+
+		storageClass, err := v.validateAndGetStorageClass(scName, pvc)
+		if err != nil {
+			return err
+		}
+
+		v.log.Info(fmt.Sprintf("storageClass %s validated", storageClass.Name))
+
+		switch len(v.instance.Spec.Async.PeerClasses) {
+		case 0:
+			v.log.Info("validate using only sc provisioner")
+
+			replicationClassMatchFound := false
+
+			for _, replicationClass := range v.replClassList.Items {
+				if storageClass.Provisioner == replicationClass.Spec.Provisioner {
+					v.volRepPVCs = append(v.volRepPVCs, *pvc)
+					replicationClassMatchFound = true
+
+					break
+				}
+			}
+
+			if !replicationClassMatchFound {
+				v.volSyncPVCs = append(v.volSyncPVCs, *pvc)
+			}
+		default:
+			v.log.Info("filtering using peerClasses")
+
+			replicationClass, err := v.findReplicationClassUsingPeerClasses(peerClasses, storageClass, pvc)
+			if err != nil {
+				return err
+			}
+
+			if replicationClass == nil && !v.instance.Spec.VolSync.Disabled {
+				// ignoring volsnapClass till we store it in the future
+				if _, err := v.findVolSnapClass(storageClass, pvc); err != nil {
+					return err
+				}
+			}
 		}
 	}
 
 	v.log.Info(fmt.Sprintf("Found %d PVCs targeted for VolRep and %d targeted for VolSync",
 		len(v.volRepPVCs), len(v.volSyncPVCs)))
 
+	if !(len(pvcList.Items) == (len(v.volRepPVCs) + len(v.volSyncPVCs))) {
+		return fmt.Errorf("no PVCs are procted")
+	}
+
+	return nil
+}
+
+func (v *VRGInstance) findReplicationClassUsingPeerClasses(peerClasses []ramendrv1alpha1.PeerClass,
+	storageClass *storagev1.StorageClass, pvc *corev1.PersistentVolumeClaim) (
+	*volrep.VolumeReplicationClass, error,
+) {
+	scName := pvc.Spec.StorageClassName
+
+	// this check is needed because we use storageID to compare with other VRC or VSC
+	if _, ok := storageClass.Labels[StorageIDLabel]; !ok {
+		msg := fmt.Sprintf("storageID label not found in storageClass for PVC %s", pvc.Name)
+		v.updatePVCDataReadyCondition(pvc.Namespace, pvc.Name, VRGConditionReasonStorageIDNotFound, msg)
+
+		return nil, fmt.Errorf(msg)
+	}
+
+	peerClass := v.findPeerClassMatchingSC(peerClasses, *scName)
+	if peerClass == nil {
+		msg := fmt.Sprintf("peerClass matching storageClass %s not found for async PVC", *scName)
+		v.updatePVCDataReadyCondition(pvc.Namespace, pvc.Name, VRGConditionReasonPeerClassNotFound, msg)
+
+		return nil, fmt.Errorf(msg)
+	}
+
+	replicationClass := v.findMatchingReplicationClass(peerClass, storageClass, pvc)
+
+	return replicationClass, nil
+}
+
+func (v *VRGInstance) findVolSnapClass(storageClass *storagev1.StorageClass,
```

Comment:
The name of the function doesn't tell me that it may add the PVC to `volSyncPVCs`